src/AdvancedMH.jl

```diff
@@ -0,0 +1,135 @@
+module AdvancedMH
+
+# Import the relevant libraries.
+using Reexport
+using Turing.Interface
+@reexport using Distributions
+using Random
+
+# Import specific functions and types to use or overload.
+import Distributions: VariateForm, ValueSupport, variate_form, value_support, Sampleable
+import MCMCChains: Chains
+import Turing.Interface: step!, AbstractSampler, AbstractTransition, transition_type
+
+# Exports
+export MetropolisHastings, DensityModel, sample
+
+# Define a sampler type.
+struct MetropolisHastings{T} <: AbstractSampler 
+    init_θ :: T
+    proposal :: Function
+end
+
+# Default constructors.
+MetropolisHastings(init_θ::T) where T<:Real = MetropolisHastings{T}(init_θ, x -> Normal(x, 1.0))
+MetropolisHastings(init_θ::T) where T<:Vector{<:Real} = MetropolisHastings{T}(init_θ, x -> MvNormal(x, 1.0))
+
+# Define a model type. Stores the log density function and the data to 
+# evaluate the log density on.
+struct DensityModel{V<:VariateForm, S<:ValueSupport, T} <: Sampleable{V, S}
+    π :: Function
+    data :: T
```

Comment:
Maybe 

- consider removing `data`, and make `π` should be a self-contained `logpdf` (i.e. `π` ∝ prior x likelihood)
- change `π` to `ℓπ` (i.e. `logdensity`)

Reply:
Also, please consider making the type of `ℓπ` parametric for type stability reasons.